docs/DesignDocs/edl_security_and_portability.md

```diff
@@ -0,0 +1,109 @@
+EDL security and portability enhancements
+=========================================
+
+Motivation
+----------
+
+This document proposes EDL enhancements that will:
+
+- Improve portability to TrustZone
+- Mitigate security vulnerabilities
+
+This feature is needed by TrustZone to ensure that applications developed with
+EDL will work on TrustZone platforms. Currently the use of host shared memory
+is problematic for TrustZone, which utilizes a different shared memory model.
+SGX *implicitly* shares *all* host memory with the enclave, whereas, TrustZone
+*explicitly* shares *some* host memory with the enclave. Therefore we
+distinguish two models.
+
+- **Implicit-all** -- implicitly shares all host memory (SGX)
+- **Explicit-some** -- explicitly shares some host memory (TrustZone)
+
+In TrustZone, developers may explicitly share designated memory regions. It
+is not possible to share all host memory (stack, data, heap, etc.) as it is
+in SGX.
+
+The feature is also needed to protect developers from security vulnerabilities
+caused by coding errors. Of particular concern are two scenarios:
+
+1. Trusted code may unwittingly write secrets onto host memory.
+
+2. Attackers may pass pointers to trusted memory where the trusted
+   code expects to write to host memory, forcing trusted memory overwrites.
+
+To prevent these kinds of errors, the stub routines must perform full
+serialization and deserialization across the trust boundary so that:
+
+- ECALL function parameters fall entirely within trusted memory.
+- OCALL function parameters fall entirely within untrusted memory.
+
+Most importantly, it should not be possible to *implicitly* pass untrusted
+memory to an ECALL function implementation. The stub routines must copy
+all untrusted memory to trusted memory buffers prior to dispatching the
+ECALL.
+
+There are currently three ways that EDL *implicitly* transfers untrusted memory.
+
+1. Through ECALLs utilizing foreign structures, which may contain pointers to
+   untrusted memory.
+
+2. Through ECALLs utilizing local structures, which are not sufficiently
+   annotated for deep copy.
+
+3. Through OCALLs that return pointers.
+
+All three of these subtly expose untrusted memory to the enclave developer.
+
+To mitigate all three of these, the **edger8r** tool should impose three
+constraints *by default*.
+
+1. Reject foreign structures.
+2. Reject under-annotated local structures.
+3. Reject pointer return values.
```

Comment:
The other way to address 3 would be to propose extending EDL to allow annotating pointer return values.

Reply:
I considered this but I thought EDL did not support function-level attribute lists.

@Andy: what do you think?